debug: use the new metrics stream in debug command

dnephin opened the pull request:
Related to #10320

Tested by running a dev agent and collecting a `consul debug` archive. All the metrics are indent formatted in the `metrics.json`.

TODO:
* [x] changelog
* [x] manual test to see what the metrics look like in the debug archive
* [ ] add a test for the API endpoint
* [ ] verify and improve the assertions in the debug CLI test for the metrics json file.

dhiaayachi commented on agent/agent_endpoint.go:
```diff
@@ -158,6 +159,54 @@ func (s *HTTPHandlers) AgentMetrics(resp http.ResponseWriter, req *http.Request)
+}
+
+func (m metricsEncoder) Encode(summary interface{}) error {
+	if err := m.encoder.Encode(summary); err != nil {
```
If I understand right, the encoder is json encoding but later we are decoding the json and re encoding it, I think we should pass the file encoder directly to the stream or use a byte encoder here and encode the json only once later

dnephin replied:
Ya, we could write it directly to the file with an `io.Copy`. I did it this way to match what we were doing before.

I think if we marshal it here with indent, then we don't need to unmarshal/marshal it again in the CLI, we can write to disk. I think this one does need to be JSON though, in case someone else is using the API directly and not using the CLI.

dhiaayachi replied:
yes right I forgot that those are public APIs

dhiaayachi commented on command/debug/debug.go:
```diff
@@ -515,20 +515,33 @@ func (c *cmd) captureLogs(timestampDir string) error {
+		if err := decoder.Decode(&raw); err != nil {
+			return fmt.Errorf("failed to decode metrics: %w", err)
+		}
+		if err := encoder.Encode(raw); err != nil {
```
There is an open issue [7872](https://github.com/golang/go/issues/7872) about the json encoder buffering the whole file and not flushing until close time. That mean that we will not flush to disk before close, not sure if it's an issue in case of app crash debug

dnephin replied:
In this case I think it is safe because the API returns multiple JSON objects. I believe the encoder only buffers until it reaches an object, so this will write each one separately. Each call to `Encode` will reach a full JSON object, and write it to the file, then the loop will continue to wait for the next full JSON object.

This can be replaced with an `io.Copy` though, if we  make the other change.

dhiaayachi commented on command/debug/debug.go:
```diff
@@ -433,12 +425,20 @@ func (c *cmd) captureLongRunning() error {
-	// Capture logs
 	if c.configuredTarget("logs") {
 		g.Go(func() error {
 			return c.captureLogs(timestampDir)
```
We should definitely do something similar to logs it will be much cleaner, but probably in a separate PR

dhiaayachi commented on agent/agent_endpoint.go:
```diff
@@ -158,6 +159,54 @@ func (s *HTTPHandlers) AgentMetrics(resp http.ResponseWriter, req *http.Request)
+		m.logger.Error("failed to encode metrics summary", "error", err)
+		return err
+	}
+	m.flusher.Flush()
```
The flush here is for each metric data we receive, I don't know how much this is memory consuming on the json encoder side

dnephin replied:
The PR that added this is: https://github.com/armon/go-metrics/pull/125

The intervals end every 10s, and it writes a full summary, which is a pretty large object. I think flushing once per summary is about right.

dhiaayachi reviewed (commented):
I saw that this touch the debug part and I took a look, I really like how the stream API work. It make it easier to handle long running requests. As commented below I think we should also change the log API to work the same way. 
I also had few comments/questions

dnephin commented:
I think the TODOs are resolved, and this should be ready for review

eculver commented on agent/agent_endpoint.go:
```diff
@@ -166,6 +167,55 @@ func (s *HTTPHandlers) AgentMetrics(resp http.ResponseWriter, req *http.Request)
+	if err != nil {
+		return nil, err
+	}
+	if rule != nil && rule.AgentRead(s.agent.config.NodeName, nil) != acl.Allow {
```
If the `rule` here is actually `nil`, the ACL check is skipped. Is there any case where this could actually happen? Would it make sense to add the `nil` check separately?

dnephin replied:
Excellent question! This `rule != nil` has given me pause as well, and it's something I'd like to change. But I believe it is correct given the way the code works today. I think basically everywhere in the code base that checks ACLs does this very same check (`git grep 'rule != nil'` returns quite a few lines that look exactly like this).

The reason can be seen here: https://github.com/hashicorp/consul/blob/v1.10.0/agent/consul/acl.go#L1187-L1190

When ACLs are disabled we return a nil `acl.Authorizer` (`rule` in this code is an authorizer, I believe the name comes from the old ACL system, and we probably just left the variable names alone when we updated the ACL system).

I think it would be great to change `ResolveToken` to have it return the [AllowAll](https://github.com/hashicorp/consul/blob/v1.10.0/acl/static_authorizer.go#L229-L232) `Authorizer` instead of `nil`. That would remove the need for a `nil` check and I think would make the code less scary to read.

I think that's a change we could make pretty easily in another PR, and see what others think.

eculver reviewed (commented):
Looks good. I just had one question about an edge-case. Also, can we add a couple tests for this new endpoint?

dnephin commented:
> can we add a couple tests for this new endpoint?

Ya, good idea. The change itself is covered by the tests for the CLI, although I could probably improve the checks in that test as well. The underlying behaviour is tested in the `go-metrics` library, but a test of the Consul integration wouldn't hurt.

dnephin commented:
Ok, I've rebased this PR and added a test for the HTTP endpoint, and improved the test coverage of the CLI.

I think this is ready for another review.

eculver reviewed (approved):
software of the highest quality.. plus tests!

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/416873.
